Add support for kubernetes provider to recognize namespace level defaults

vjsamuel opened the pull request:
This PR allows namespace level defaults to be configured on the namespace object for either service or pod hints autodiscover. This will need:

```
add_resource_metadata.namespace.enabled: true
```

to be set on the kubernetes autodiscover provider. 

This PR has a known gap where a pod is deployed before the namespace annotations are added. In this case, autodiscover wouldnt be able to add in the defaults. To fix that behavior we would need to ensure that we handle resync events as updates. Currently we ignore resync events as the resource version remains the same.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

blakerouse commented:
We have a template that is generally used when creating a PR, could you update this PR to fill out all those sections.

Also did (or could) you file a bug for the behavior your trying to solve and can you provide steps on how to test this PR?

exekias commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -188,10 +190,18 @@ func (p *pod) GenerateHints(event bus.Event) bus.Event {
+
+	// Fall back to defaults on the namespace if there were no hints on the pods
+	if len(hints) == 0 {
+		if rawAnn, ok := kubeMeta["defaults"]; ok {
```
It sounds to me that we should do a mix of namespace hints + pod specific hints? Example:

I may have a setting to do json decode for all pods, but for one specific pod in the namespace I may want to tune how multiline goes, without affecting the JSON parsing?

vjsamuel commented:
I have added changes such that annotations that are missing on the pod/service are defaulted with the help of namespace level annotations. i have added test cases for the same. @exekias PTAL

exekias commented on libbeat/autodiscover/providers/kubernetes/kubernetes.go:
```diff
@@ -142,8 +142,11 @@ func (p *Provider) publish(event bus.Event) {
+		if config := p.builders.GetConfig(e); config != nil {
 			event["config"] = config
+		} else {
+
```
empty else?

exekias commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -296,6 +312,18 @@ func (p *pod) emitEvents(pod *kubernetes.Pod, flag string, containers []kubernet
+					for k, v := range namespace.GetAnnotations() {
+						safemapstr.Put(defaults, k, v)
+					}
+					kubemeta["defaults"] = defaults
```
I don't think we need to create a new key for this. metadata package is already able to put namespace annotations under `kubernetes.namespace.annotations` right? We can add that to the `kubemeta` object, in the same way we always add pod `annotations`.

vjsamuel replied:
onl in 8.0 we will have `namespace` as a `MapStr`. until then it would be `namespace_annotations`

vjsamuel replied:
thoughts?

exekias replied:
good catch, that's true! I think we should do the same here

exekias commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -159,15 +161,27 @@ func (p *pod) GenerateHints(event bus.Event) bus.Event {
+		}
+
+		// Look at all the namespace level default annotations and do a merge with priority going to the pod annotations.
+		if rawNsAnn, ok := kubeMeta["defaults"]; ok {
```
we would be getting `namespace.annotations` here

exekias commented:
Thank you for the changes, I think this is looking good! let's have the docs in and it's ready to go :tada:

vjsamuel commented:
@exekias added docs.

exekias commented:
ok to test

exekias commented:
How to test this:

Deploy Filebeat (or Metricbeat, but Filebeat makes more sense here) in kubernetes, enable hints based autodiscover with `add_resource_metadata.namespace.enabled: true`. Now adding namespace hints should apply to all pods in the namespace.

For instance, adding this annotation to the namespace:

```
co.elastic.logs/enabled: false
```

Should result on Filebeat stop retrieving logs from the pods in that namespace, while the others should still be retrieved. Also we should test that this can be overridden per pod, for instance, disable collection for a whole namespace, but enable it for one of the pods only
